### PR TITLE
reserved words with brackets in autocomplete

### PR DESCRIPTION
**Pull request by abist**
This fixes [@473](https://github.com/Microsoft/vscode-mssql/issues/473).

### CODE REVIEW

File: src/Microsoft.SqlTools.ServiceLayer/LanguageServices/AutoCompleteHelper.cs [hunk 1]
```diff
@@ -343,6 +343,14 @@ internal static CompletionItem[] EmptyCompletionList
         }
 
         /// <summary>
+        /// Retrieves the reserved word list 
+        /// </summary>
+        public static string[] GetReservedWordList()
+        {
+            return DefaultCompletionText;
```
Comment (kburtram) [inline src/Microsoft.SqlTools.ServiceLayer/LanguageServices/AutoCompleteHelper.cs]: >DefaultCompletionText [](start = 19, length = 21)

I think we can just make DefaultCompletionText internal instead of adding this method.

Comment (kburtram) [inline src/Microsoft.SqlTools.ServiceLayer/LanguageServices/AutoCompleteHelper.cs]: If you don't want to add a separate assessor, please  add it as a property similar to EmptyCompletionList above.

---
In reply to: [92691614](https://github.com/Microsoft/sqltoolsservice/pull/191#discussion_r92691614) [](ancestors = 92691614)

File: src/Microsoft.SqlTools.ServiceLayer/LanguageServices/Completion/SqlCompletionItem.cs [hunk 2]
```diff
@@ -45,7 +46,7 @@ private void Init()
         {
             InsertText = GetCompletionItemInsertName();
             Label = DeclarationTitle;
-            if (StartsWithBracket(TokenText))
+            if (StartsWithBracket(TokenText) || IsReservedWord(InsertText))
```
Comment (kburtram) [inline src/Microsoft.SqlTools.ServiceLayer/LanguageServices/Completion/SqlCompletionItem.cs]: >IsReservedWord [](start = 48, length = 14)

have you verified we don't add '[' when autocompleted keywords outside the column or table completion context...e.g., we don't escape SELECT if it's typed by its self.

**Comment by kburtram** [comment]
:clock1:

File: src/Microsoft.SqlTools.ServiceLayer/LanguageServices/Completion/SqlCompletionItem.cs (hunk 2 above)
Comment (abist) [inline src/Microsoft.SqlTools.ServiceLayer/LanguageServices/Completion/SqlCompletionItem.cs]: Yes, I did and it works as intended.

File: src/Microsoft.SqlTools.ServiceLayer/LanguageServices/AutoCompleteHelper.cs (hunk 1 above)
Comment (abist) [inline src/Microsoft.SqlTools.ServiceLayer/LanguageServices/AutoCompleteHelper.cs]: Changed it

File: src/Microsoft.SqlTools.ServiceLayer/LanguageServices/Completion/SqlCompletionItem.cs [hunk 3]
```diff
@@ -202,5 +203,12 @@ private string WithBracket(string text)
                 return text;
             }
         }
+
+        private bool IsReservedWord(string text)
```
Comment (abist) [inline src/Microsoft.SqlTools.ServiceLayer/LanguageServices/Completion/SqlCompletionItem.cs]: Done.

File: src/Microsoft.SqlTools.ServiceLayer/LanguageServices/Completion/SqlCompletionItem.cs [hunk 4]
```diff
@@ -45,7 +46,7 @@ private void Init()
         {
             InsertText = GetCompletionItemInsertName();
             Label = DeclarationTitle;
-            if (StartsWithBracket(TokenText))
+            if (StartsWithBracket(TokenText) || AutoCompleteHelper.IsReservedWord(InsertText))
             {
                 Label = WithBracket(Label);
                 InsertText = WithBracket(InsertText);
```
Comment (llali) [inline src/Microsoft.SqlTools.ServiceLayer/LanguageServices/Completion/SqlCompletionItem.cs]: I think for your scenario you should only add the bracket to insert text not the label. because now user see everything in the list without bracket except for objects with reserved word name

File: src/Microsoft.SqlTools.ServiceLayer/LanguageServices/Completion/SqlCompletionItem.cs [hunk 5]
```diff
@@ -45,7 +46,7 @@ private void Init()
         {
             InsertText = GetCompletionItemInsertName();
             Label = DeclarationTitle;
-            if (StartsWithBracket(TokenText))
+            if (StartsWithBracket(TokenText) || AutoCompleteHelper.IsReservedWord(InsertText))
             {
```
Comment (llali) [inline src/Microsoft.SqlTools.ServiceLayer/LanguageServices/Completion/SqlCompletionItem.cs]: also I don't see any test tp verify new scenarios. can you please add tests?

**Comment by llali** [comment]
:clock1:

File: src/Microsoft.SqlTools.ServiceLayer/LanguageServices/Completion/SqlCompletionItem.cs (hunk 5 above)
Comment (abist) [inline src/Microsoft.SqlTools.ServiceLayer/LanguageServices/Completion/SqlCompletionItem.cs]: sure, on it

File: src/Microsoft.SqlTools.ServiceLayer/LanguageServices/Completion/SqlCompletionItem.cs (hunk 4 above)
Comment (abist) [inline src/Microsoft.SqlTools.ServiceLayer/LanguageServices/Completion/SqlCompletionItem.cs]: I was trying to be consistent with other words that start with brackets. Is it better to remove it from the label?

File: src/Microsoft.SqlTools.ServiceLayer/LanguageServices/AutoCompleteHelper.cs [hunk 6]
```diff
@@ -343,6 +343,27 @@ internal static CompletionItem[] EmptyCompletionList
         }
 
         /// <summary>
+        /// Retrieves the reserved word list 
+        /// </summary>
+        internal static string[] GetReservedWordList
```
Comment (kburtram) [inline src/Microsoft.SqlTools.ServiceLayer/LanguageServices/AutoCompleteHelper.cs]: > internal static string[] GetReservedWordList [](start = 7, length = 45)

this property doesn't appear to be used anywhere now since IsReserveredWord was moved into this class...if it's no longer needed can we delete it?

Comment (abist) [inline src/Microsoft.SqlTools.ServiceLayer/LanguageServices/AutoCompleteHelper.cs]: agreed, removed it

**Comment by penwon** [comment]
Is it possible to disable escaping for temp table names? For example in query like: select * from #MyTable. #MyTable transformed to [#MyTable].

File: test/Microsoft.SqlTools.ServiceLayer.Test/LanguageServer/SqlCompletionItemTests.cs [hunk 7]
```diff
@@ -13,6 +13,304 @@ namespace Microsoft.SqlTools.ServiceLayer.Test.LanguageServer
 {
     public class SqlCompletionItemTests
     {
+        private static readonly string[] ReservedWords = new string[]
+        {
+            "all",
```
Comment (kburtram) [inline test/Microsoft.SqlTools.ServiceLayer.Test/LanguageServer/SqlCompletionItemTests.cs]: I thought we already had this list in the language server code.  Perhaps it makes more sense to have the test access the list from there rather than duplicate it into the test code.

**Comment by abist** [comment]
@penwon Thanks for the suggestion. Opened a new issue for this here - [#692](https://github.com/Microsoft/vscode-mssql/issues/692)
